### ISSUE TITLE
WebAssembly: Unable to obtain funcref from pointer

### ISSUE THREAD

**Issue by nokotan** [llvm/llvm-project#53789]
In a WebAssembly assembly code, unable to obtain funcref from function pointer.

```wat
.global funcptr_to_funcref
.global __indirect_function_table
.table_type __indirect_function_table, funcref

funcptr_to_funcref:
  .functype funcptr_to_funcref (i32) -> (funcref)
  local.get 0
  table.get __indirect_function_table
  end_function
```

```log
popped externref, expected funcref
```

Table type of __indirect_function_table is funcref, it is expected that table.get will push a corresponding funcref into the stack.

**Comment by sbc100** [comment]
@pmatos @aardappel. 

Should we add some way to disabled the type checker, or downgrade it to warnings, since I imagine this would actually work fine if that type checker wasn't confused (i.e. you would get a valid object file).

**Comment by aardappel** [comment]
There is already a flag to turn off the type checker, but really, the better action is to enhance the type checker to understand this type.

The type checker should be considered a part of Wasm asm (hah?) from now. Given that validation is part of the Wasm spec, and that writing arbitrary instruction sequences as in other assembly languages doesn't make sense for Wasm, asm should be type checked much like .wat is.

**Comment by sbc100** [comment]
Sure, I'm certainly not saying this doesn't need fixing, just wondering if we can unblock OP while a fix is being worked on. 

What is the flag?

**Comment by aardappel** [comment]
`-no-type-check` (to at least `llvm-mc` I think)

**Comment by nokotan** [comment]
I'm quite new to LLVM DAG, `table.get` instruction seems to be defined twice; externref and funcref.

https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/WebAssembly/WebAssemblyInstrTable.td

Adding table.* pattern resolved this this error for me, I'll create new PR that includes this patch.

https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/WebAssembly/AsmParser/WebAssemblyAsmTypeCheck.cpp#L204

```cpp
  if (Name == "table.get") {
    if (getTable(ErrorLoc, Inst, Type))
      return true;
    if (popType(ErrorLoc, wasm::ValType::I32))
      return true;
    Stack.push_back(Type);
  } else if (Name == "table.set") {
    if (getTable(ErrorLoc, Inst, Type))
      return true;
    if (popType(ErrorLoc, Type))
      return true;
    if (popType(ErrorLoc, wasm::ValType::I32))
      return true;
  }
```

**Comment by aardappel** [comment]
We generally have our tablegen set up such that all instructions are generated for every possible type they operate on.

**Comment by pmatos** [comment]
I am sorry I missed this bug earlier on and just noticed this as @sbc100 marked me as a reviewer on llvm phab. I will take a look at the bug and patch a bit later on today.

**Comment by pmatos** [comment]
I have left some comments in phab. Thanks for the patch.

**Comment by pmatos** [comment]
This seems to have successfully landed. Closing.
Thanks @nokotan for the patch.
